File: src/main/java/me/lightspeed7/mongofs/writing/CountingOutputStream.java
package me.lightspeed7.mongofs.writing;

import java.io.IOException;
import java.io.OutputStream;

import me.lightspeed7.mongofs.MongoFile;
import me.lightspeed7.mongofs.common.MongoFileConstants;

public class CountingOutputStream extends OutputStream {

    long count = 0;
    private MongoFileConstants key;
    private MongoFile inputFile;
    private OutputStream out;

    public CountingOutputStream(MongoFileConstants key, MongoFile inputFile, OutputStream out) {

        this.key = key;
        this.inputFile = inputFile;
        this.out = out;
    }

    @Override
    public void write(int b)
            throws IOException {

        out.write(b);
        ++count;
    }

    @Override
    public void write(byte[] b, int off, int len)
            throws IOException {

        out.write(b, off, len);
        count += len;
    }

    @Override
    public void close()
            throws IOException {

        out.close();

        inputFile.put(key.toString(), count);
        inputFile.save();
    }
}

File: reactive/src/main/scala/me/lightspeed7/mongofs/stream/CountingOutputStream.scala
package me.lightspeed7.mongofs.stream

import java.io.OutputStream
import java.io.IOException
import me.lightspeed7.mongofs._
import me.lightspeed7.mongofs.url.MongoFileUrl
import org.joda.time.DateTime
import java.util.concurrent.atomic.AtomicLong
import scala.concurrent.Future
import java.util.concurrent.atomic.AtomicInteger

class CountingOutputStream(stats: MongoFileStatistics, out: OutputStream)(implicit val store: MongoFileStore) extends OutputStream {

  @throws(classOf[IOException])
  def write(b: Int): Unit = { stats.incrementTotal(1); out.write(b) }

  @throws(classOf[IOException])
  override def write(b: Array[Byte]): Unit = { stats.incrementTotal(b.length); out.write(b) }

  @throws(classOf[IOException])
  override def write(b: Array[Byte], o: Int, l: Int): Unit = { stats.incrementTotal(l - o); out.write(b, o, l) }

}